fastjsonschema/generator.py

```diff
@@ -245,8 +245,27 @@ def exc(self, msg, *args, rule=None):
         Short-cut for creating raising exception in the code.
         """
         msg = 'raise JsonSchemaValueException("'+msg+'", value={variable}, name="{name}", definition={definition}, rule={rule})'
-        definition_rule = self.e(self._definition.get(rule) if isinstance(self._definition, dict) else None)
-        self.l(msg, *args, definition=repr(self._definition), rule=repr(rule), definition_rule=definition_rule)
+        definition = self._expand_refs(self._definition)
+        definition_rule = self.e(definition.get(rule) if isinstance(definition, dict) else None)
+        self.l(msg, *args, definition=repr(definition), rule=repr(rule), definition_rule=definition_rule)
+
+    def _expand_refs(self, definition):
+        if isinstance(definition, list):
+            return [self._expand_refs(v) for v in definition]
+        if not isinstance(definition, dict):
+            return definition
+        if "$ref" in definition:
+            with self._resolver.resolving(definition["$ref"]) as schema:
+                return schema
+        if "properties" in definition:
+            return {
+                **definition,
```

Comment:
Shouldn't definition here also recursively expand refs?

Reply:
Tried to look around and found scenario here: https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties

```
{
  "type": "object",
  "properties": {
    "street_address": { "type": "string" },
    "city": { "type": "string" },
    "state": { "type": "string" },
    "type": { "enum": ["residential", "business"] }
  },
  "required": ["street_address", "city", "state", "type"],

  "if": {
    "type": "object",
    "properties": {
      "type": { "const": "business" }
    },
    "required": ["type"]
  },
  "then": {
    "properties": {
      "department": { "type": "string" }
    }
  },

  "unevaluatedProperties": false
}
```

In other words, if there are properties, there could be also if/then which can also contain refs. So I think that deserves tests on its own and do expansion here as well. Do you agree or maybe I have it wrong and this is enough? Maybe in either case test like that would be helpful so we can see its not needed as well. Do you think you can do that?

Reply:
Yeah, I agree that it is a good idea to add a test case for that.

Let me add it ;)

Reply:
Thank you, appreciated!